reana_server/scheduler.py

```diff
@@ -125,67 +135,106 @@ def get_consumers(self, Consumer, channel):
+                **workflow_submission
+            )
+
+    def on_message(self, body, message):
         """On new workflow_submission event handler."""
-        workflow_submission = json.loads(workflow_submission)
-        logging.info("Received workflow: {}".format(workflow_submission))
+        workflow_submission = json.loads(body)
+        logging.info(f"Received workflow: {workflow_submission}")
+
+        workflow_submission_copy = workflow_submission.copy()
+
+        workflow_id = workflow_submission["workflow_id_or_name"]
         workflow_min_job_memory = workflow_submission.pop("min_job_memory", 0)
+
+        workflow_submission.pop("priority", None)
+        workflow_submission.pop("retry_count", None)
+
         if reana_ready(workflow_min_job_memory):
-            logging.info("Starting queued workflow: {}".format(workflow_submission))
-            workflow_submission.pop("priority", 0)
+            logging.info(f"Starting queued workflow: {workflow_id}")
             workflow_submission["status"] = "start"
+
+            retry = True
+            should_fail_on_error = True
+            started = False
+            error_message = ""
+
             try:
-                requeue = True
-                started = False
                 (
                     response,
                     http_response,
                 ) = current_rwc_api_client.api.set_workflow_status(
                     **workflow_submission
                 ).result()
                 http_response_json = http_response.json()
-                if http_response.status_code == 200:
-                    started = True
-                    logging.info(
-                        f'Workflow {http_response_json["workflow_id"]} successfully started.'
-                    )
-                else:
-                    logging.error(
-                        f"RWC returned an unexpected status code:\n {http_response_json}"
-                    )
+                started = True
+                logging.info(
+                    f'Workflow {http_response_json["workflow_id"]} successfully started.'
+                )
 
             except HTTPBadGateway as api_e:
+                retry = True
+                should_fail_on_error = True
                 logging.error(
                     "Workflow failed to start because RWC got an error while calling"
                     f"an external service (i.e. DB):\n {api_e}",
                     exc_info=True,
                 )
             except HTTPNotFound as not_found_e:
+                retry = False
+                should_fail_on_error = False
                 logging.error(
                     "Workflow failed to start because it does not exist or was deleted \n"
                     f"{not_found_e}",
                     exc_info=True,
                 )
-                requeue = False
             except HTTPConflict as e:
+                retry = False
+                should_fail_on_error = False
```

Comment:
Minor: IMO small comment to explain that in this case the message should be `acked` silently would really help here

Reply:
What about some duplication like this, makes things very clear:

```python
try:
    (
        response,
        http_response,
    ) = current_rwc_api_client.api.set_workflow_status(
        **workflow_submission
    ).result()
    http_response_json = http_response.json()
    logging.info(
        f'Workflow {http_response_json["workflow_id"]} successfully started.'
    )
    sleep(REANA_SCHEDULER_REQUEUE_SLEEP)
    message.ack()
except HTTPBadGateway as api_e:
    logging.error(
        "Workflow failed to start because RWC got an error while calling"
        f"an external service (i.e. DB):\n {api_e}",
        exc_info=True,
    )
    sleep(REANA_SCHEDULER_REQUEUE_SLEEP)
    message.reject()
    self._retry_submission(workflow_id, workflow_submission_copy)
except HTTPNotFound as not_found_e:
    logging.error(
        "Workflow failed to start because it does not exist or was deleted \n"
        f"{not_found_e}",
        exc_info=True,
    )
    sleep(REANA_SCHEDULER_REQUEUE_SLEEP)
    # if workflow is not found, we cannot retry or report an error to workflow logs
    message.reject()
except HTTPConflict as e:
    logging.error(
        f"Workflow failed to start because of duplicated message from RabbitMQ.\n {e}",
        exc_info=True,
    )
    sleep(REANA_SCHEDULER_REQUEUE_SLEEP)
    message.reject()
except Exception as e:
    logging.error(
        f"Something went wrong while calling RWC:\n {e}", exc_info=True
    )
    sleep(REANA_SCHEDULER_REQUEUE_SLEEP)
    message.reject()
    self._retry_submission(workflow_id, workflow_submission_copy)
```

My concerns with comments are that they can get outdated, and no one will change them.

Reply:
IMO something simple as this would work (haven't tested):
```python
    def on_message(self, body, message):
        """On new workflow_submission event handler."""
        workflow_submission = json.loads(body)
        logging.info(f"Received workflow: {workflow_submission}")

        workflow_min_job_memory = workflow_submission.pop("min_job_memory", 0)
        retry_count = workflow_submission.pop("retry_count", 0)

        if reana_ready(workflow_min_job_memory):
            logging.info(f"Starting queued workflow: {workflow_submission}")
            workflow_submission["status"] = "start"
            priority = workflow_submission.pop("priority", 0)

            requeue = True
            started = False

            try:
                (
                    response,
                    http_response,
                ) = current_rwc_api_client.api.set_workflow_status(
                    **workflow_submission
                ).result()
                http_response_json = http_response.json()
                if http_response.status_code == 200:
                    started = True
                    logging.info(
                        f'Workflow {http_response_json["workflow_id"]} successfully started.'
                    )
                else:
                    logging.error(
                        f"RWC returned an unexpected status code:\n"
                        f"{http_response_json}"
                    )

            except HTTPBadGateway as api_e:
                logging.error(
                    "Workflow failed to start because RWC got an error while calling"
                    f"an external service (i.e. DB):\n {api_e}",
                    exc_info=True,
                )
            except HTTPNotFound as not_found_e:
                logging.error(
                    "Workflow failed to start because it does not exist or was deleted \n"
                    f"{not_found_e}",
                    exc_info=True,
                )
                requeue = False
            except HTTPConflict as e:
                logging.error(
                    f"Workflow failed to start because of Rabbit MQ message conflict.\n {e}",
                    exc_info=True,
                )
                requeue = False
            except Exception as e:
                logging.error(
                    f"Something went wrong while calling RWC:\n{e}", exc_info=True
                )
            finally:
                sleep(REANA_SCHEDULER_REQUEUE_SLEEP)
                if not started and requeue:
                    message.reject()
                    self._retry_submission(
                        **workflow_submission,
                        priority=priority,
                        min_job_memory=workflow_min_job_memory,
                    )
                else:
                    message.ack()
        else:
            logging.info(
                f'REANA not ready to run workflow {workflow_submission["workflow_id_or_name"]}.'
            )
            sleep(REANA_SCHEDULER_REQUEUE_SLEEP)
            message.reject()
            self._retry_submission(
                **workflow_submission,
                min_job_memory=workflow_min_job_memory,
                retry_count=retry_count,
            )
```